movingpandas/trajectory_generalizer.py

```diff
@@ -203,3 +204,49 @@ def _generalize_traj(self, traj, tolerance):
         new_df = traj.df.iloc[keep_rows]
         new_traj = Trajectory(new_df, traj.id)
         return new_traj
+
+
+class TopDownTimeRatioGeneralizer(TrajectoryGeneralizer):
+    """
+    Generalizes using Top-Down Time Ratio algorithm proposed by Meratnia and De By (DOI: 10.1007/978-3-540-24741-8_44).
+
+    Points are projected on the line that connects the starting point SP with the end point EP of the trajectory. These projections are
+    calculated based on the ratios de and di that correspond to the travel time from SP to P and from P to EP respectively. Iterativelly,
+    when the furthest point P is found (granted its distance is greater that the tolerance), that point is kept and the process is
+    repeated for the two trajectories that are defined as follows: (SP,P) and (P,EP). The process is stopped when any indivudiual 
+    subtrajectory is small enough (len<=2). 
+
+    tolerance : float
+        Distance tolerance (distance returned by shapely Point.distance function)
+
+    Examples
+    --------
+
+    >>> mpd.TopDownTimeRatioGeneralizer(traj).generalize(tolerance=1.0)
+    """
+
+    def _generalize_traj(self, traj, tolerance):
+        return Trajectory(self.td_tr(traj.df.copy(), tolerance), traj.id)
+        
+    def td_tr(self, df, tolerance):
+        # print(df)
+        if len(df)<=2:
+            return df
+        else:
+            de = (df.index.max().to_pydatetime() - df.index.min().to_pydatetime()).total_seconds()
+            
+            dx = df.geometry.iloc[-1].x - df.geometry.iloc[0].x
+            dy = df.geometry.iloc[-1].y - df.geometry.iloc[0].y
+
+            # distances for each point and the calulated one based on it and start
+            dists = df.apply(lambda rec: self._dist_from_calced(rec, df.index.min().to_pydatetime(), df.geometry.iloc[0], de, dx, dy), axis=1) 
+
+            if dists.max()>tolerance:
+                return pd.concat([self.td_tr(df.iloc[:df.index.get_loc(dists.idxmax())], tolerance), self.td_tr(df.iloc[df.index.get_loc(dists.idxmax()):], tolerance)])
+            else:
+                return df.iloc[[0,-1]]
+    
+    def _dist_from_calced(self, rec, start_t, start_geom, de, dx, dy):
+        di = (rec.name - start_t).seconds
```

Comment:
Should that also be total_seconds()?

Reply:
Yes, exactly. That was why the point 3,0 was added. I fixed that bug as well as an indexing mistake (forgot an important +1 to the left side of the concat - both parts need to be inclusive of both start and end). The result is now identical to the dp algorithm, which makes sense in a way.